_posts/2022-01-21-safe-releases.md

```diff
@@ -0,0 +1,95 @@
+
+While there are many tools at our disposal to solve some of these problems, I believe a solid understanding of the basics helps us reason about releases and avoid the biggest problems in the first place.
+
+Below I will outline the release cycle for many modern apps (with a focus on Heroku/AWS and Ruby/Javascript), as well as some guidelines to plan graceful releases and migrations.
+
+![Typical release cycle](/images/posts/2022-01-21-safe-releases/release-process.png)
+
+## Release Cycle Steps 1 & 2
+
+Many modern apps follow this 4-step deploy cycle (of course, your setup may vary).
+
+Steps one and two are pretty safe for most apps. You click a deploy button or execute a CLI command (perhaps from your CI tool) which kicks off the build process for your new release. In our apps, that executes a `rails assets:precompile` run which ends up compiling our Javascript assets with webpacker.
+
+## Danger Zone 1 - Executing Release Commands
+
+These are oftentimes data setup steps that need to be in place before new code is running. Things like:
+Database migrations
+Uploading build assets to external sources
+Preparing new environment configurations
+
+This is a danger zone because your data changes need to be backwards-compatible with both versions of code which may be running against your environment. Before the next step, your old code is still being run. After the deployment, your new code will be active.
+
+## Danger Zone 2 - Code Deploy Phase
+
+Finally, your new code is pushed to servers. This can happen in one go or it can be controlled with architectures like blue-green deployments. This is also a danger zone because the new code will fail if your release commands did not prepare the environment properly.
+
+
+## Additional Considerations
+
+Besides these two hot-spots, there are a few other things you should consider before kicking off any production deployments:
+
+### Is your release affecting a distributed system?
+
+Releasing changes to a single application is one thing, but if your system interacts with others and your changes could affect those interactions, you'll want to take special care.
+
+Below I outline a general process which can also be applied to multi-system releases, while noting that there will generally be many more releases involved in order to gracefully transition each system safely while it is still interacting with other apps.
+
+### Does your release contain long-running data migrations?
+
+One common request in a software business is to release a feature and prefill a bunch of data. Since Rails migrations support full Ruby and Rails code, it's tempting to just write a loop in there and populate a bunch of records.
```

Comment:
Great opportunity to reference the Rails Doctrine of "Provide sharp knives"

https://rubyonrails.org/doctrine#:~:text=There's%20nothing%20programmatically%20in%20Ruby,use%20spoons%20to%20slice%20tomatoes.

It's one of my fav concepts in Rails and support for full Ruby and Rails code and context in migrations is an example of that - You have the sharp knife, but make sure you are not cutting yourself with it.